src/Qir/Runtime/lib/Simulators/OpenSystemsSimulator.cpp

```diff
@@ -0,0 +1,292 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include <assert.h>
+#include <bitset>
+#include <complex>
+#include <iostream>
+#include <memory>
+#include <vector>
+
+#include "QirRuntimeApi_I.hpp"
+#include "QSharpSimApi_I.hpp"
+#include "SimFactory.hpp"
+
+using namespace std;
+typedef long long int i64;
+
+// Define the API with the open systems simulator.
+extern "C"
+{
+    // NB: We disable clang-tidy rules for case conventions here, as the names
+    //     reflect the Rust naming conventions used in the opensim crate.
+    size_t init(size_t initialcapacity); // NOLINT(readability-identifier-naming)
+    i64 destroy(size_t sim_id); // NOLINT(readability-identifier-naming)
+    void dump_to_console(size_t sim_id); // NOLINT(readability-identifier-naming)
+    i64 x(size_t sim_id, size_t idx); // NOLINT(readability-identifier-naming)
+    i64 y(size_t sim_id, size_t idx); // NOLINT(readability-identifier-naming)
+    i64 z(size_t sim_id, size_t idx); // NOLINT(readability-identifier-naming)
+    i64 h(size_t sim_id, size_t idx); // NOLINT(readability-identifier-naming)
+    i64 s(size_t sim_id, size_t idx); // NOLINT(readability-identifier-naming)
+    i64 s_adj(size_t sim_id, size_t idx); // NOLINT(readability-identifier-naming)
+    i64 t(size_t sim_id, size_t idx); // NOLINT(readability-identifier-naming)
+    i64 t_adj(size_t sim_id, size_t idx); // NOLINT(readability-identifier-naming)
+    i64 cnot(size_t sim_id, size_t idx_control, size_t idx_target); // NOLINT(readability-identifier-naming)
+    i64 m(size_t sim_id, size_t idx, size_t* result_out); // NOLINT(readability-identifier-naming)
+    const char* lasterr(); // NOLINT(readability-identifier-naming)
+    const char* get_noise_model(size_t sim_id); // NOLINT(readability-identifier-naming)
+    i64 set_noise_model(size_t sim_id, const char* new_model); // NOLINT(readability-identifier-naming)
+    const char* get_current_state(size_t sim_id); // NOLINT(readability-identifier-naming)
+}
+
+namespace Microsoft
+{
+namespace Quantum
+{
+    // FIXME: support methods from public IDiagnostics; they currently
+    //        just throw.
+    class COpenSystemSimulator : public IRuntimeDriver, public IQuantumGateSet, public IDiagnostics
```

Comment:
Consider breaking up the class into a separate definition of the class (in the .hpp header), e.g. like in [QubitManager.hpp](https://github.com/microsoft/qsharp-runtime/blob/main/src/Qir/Runtime/public/QubitManager.hpp),
and the definition of the member functions (in the .cpp source), e.g. like in [QubitManager.cpp](https://github.com/microsoft/qsharp-runtime/blob/main/src/Qir/Runtime/lib/QIR/QubitManager.cpp).

In the definition of the class (.hpp) consider 
* listing the `public`s first, then `protected`s, then `private`s (such that the users of the class see the most used API first);
* grouping all the member variables together in the end of the class in the separate `private` section (such that the users can quickly estimate the size of the class instance and whether all the member variables are initialized in the constructors).

Reply:
This was directly copied from the full-state simulator, then modified for the new API. I think the full-state simulator has diverged fairly substantially since then, including the cleanup that you mention above.

Reply:
Added to the issue for QIR integration at https://github.com/microsoft/qsharp-runtime/issues/713; if it's alright with you, I'd like to defer that along with other QIR integration work. Thanks for raising these code quality issues, appreciate your help with the C++ side!